### PR TITLE
CB-11419 Fix waiting edge cases in integration tests

### PR DESCRIPTION
**Pull request by Bajzathd**
While WaitService waits for a certain criteria to be met (depedning on StatusChecker implementation), it should refresh its data only once per round. This way we can not encounter an edge case where a check runs on different data then the check before it, causing changes to not handled properly.

See detailed description in the commit message.

### CODE REVIEW

**Review by aszegedi — APPROVED** [review]

**Review by aszegedi — APPROVED** [review]

**Review by aszegedi — APPROVED** [review]

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitFailedChecker.java [hunk 1]
```diff
@@ -14,37 +14,44 @@
...
             LOGGER.error("Failed to get {} status or statusReason: {}", waitObject.getClass().getSimpleName(), e.getMessage(), e);
             throw new TestFailException("Failed to get " + waitObject.getClass().getSimpleName() + " status or statusReason", e);
         }
-        return false;
+    }
+
+    @Override
+    public boolean checkStatus(T waitObject) {
+        if (failed) {
+            failed = false;
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitFailedChecker.java:35]: why do you set it to false?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java [hunk 2]
```diff
@@ -28,12 +28,13 @@
         while (!timeout && !exit) {
             LOGGER.info("Waiting round {} and elapsed time {} ms", attempts, System.currentTimeMillis() - startTime);
             try {
-                success = statusChecker.checkStatus(t);
+                statusChecker.refresh(t);
             } catch (Exception ex) {
                 LOGGER.debug("Exception occurred during waiting: {}", ex.getMessage(), ex);
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java:33]: could you please change the error message to `Exception occurred during refresh`

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java [hunk 3]
```diff
@@ -28,12 +28,13 @@
...
                 failures++;
                 actual = ex;
             }
+            success = statusChecker.checkStatus(t);
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java:37]: you can do this after the failure check

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitTerminationChecker.java [hunk 4]
```diff
@@ -16,45 +13,47 @@
...
-                LOGGER.error("Cluster '{}' termination failed (status:'{}'), waiting is cancelled.", name, actualStatuses);
-                throw new TestFailException(String.format("Cluster '%s' termination failed. Status: '%s' statusReason: '%s'",
-                        name, actualStatuses, actualStatusReasons));
-            }
-            if (!waitObject.isDeleted()) {
-                return false;
-            }
         } catch (NotFoundException e) {
             LOGGER.warn("No {} found with name: {}", waitObject.getClass().getSimpleName(), name);
+            deleted = true;
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitTerminationChecker.java:25]: maybe it better if store this in the wait object

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitFailedChecker.java [hunk 5]
```diff
@@ -14,37 +14,44 @@
...
-                throw new TestFailException(String.format("Cluster '%s' has been terminated. Status: '%s' statusReason: '%s'",
-                        name, actualStatuses, actualStatusReasons));
-            }
-            if (waitObject.isInDesiredStatus()) {
-                return true;
-            }
         } catch (NotFoundException e) {
-            LOGGER.warn("No {} found with name: {}", waitObject.getClass().getSimpleName(), name, e);
+            LOGGER.warn("No {} found with name: {}", waitObject.getClass().getSimpleName(), waitObject.getName(), e);
+            failed = true;
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitFailedChecker.java:25]: is it failed or deleted? maybe it better if store in the wait object

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitTerminationChecker.java [hunk 6]
```diff
@@ -64,30 +63,14 @@ public String successMessage(T waitObject) {
...
     public Map<String, String> getStatuses(T waitObject) {
-        String name = waitObject.getName();
-        try {
-            waitObject.fetchData();
-            return waitObject.actualStatuses();
-        } catch (NotFoundException e) {
-            LOGGER.warn("No cluster found with name '{}'! It has been deleted successfully.", name, e);
-            return Map.of("status", DELETE_COMPLETED.name());
-        }
+        return deleted ? waitObject.getDesiredStatuses() : waitObject.actualStatuses();
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitTerminationChecker.java:74]: what will happen in the test if this method gives back the desires statuses? In the catch clause, we used the `DELETE_COMPLETED`

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/instance/InstanceFailedChecker.java [hunk 7]
```diff
@@ -19,48 +17,53 @@
...
-                throw new TestFailException(String.format("The '%s' has been terminated, waiting is cancelled." +
-                        " Status: '%s' statusReason: '%s'", instanceGroupName, instanceStatus, hostStatusReason));
-            }
-            if (waitObject.isInDesiredStatus()) {
-                LOGGER.info("The '{}' is in desired state (status:'{}').", instanceGroupName, instanceStatus);
-                return true;
-            }
         } catch (NoSuchElementException e) {
             LOGGER.warn("No instance group found with name '{}'", hostGroup, e);
+            failed = true;
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/instance/InstanceFailedChecker.java:29]: is it failed or deleted?

**Review by topolyai5 — CHANGES_REQUESTED** [review]

**Review by topolyai5 — APPROVED** [review]
